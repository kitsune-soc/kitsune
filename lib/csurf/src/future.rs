use crate::{CsrfHandle, CSRF_COOKIE_NAME};
use cookie::{Cookie, Expiration, SameSite};
use http::{header, HeaderValue, Response};
use pin_project_lite::pin_project;
use std::{
    future::Future,
    pin::Pin,
    task::{self, ready, Poll},
};

pin_project! {
    pub struct ResponseFuture<F> {
        #[pin]
        pub(crate) inner: F,
        pub(crate) handle: CsrfHandle,
    }
}

impl<F, E, ResBody> Future for ResponseFuture<F>
where
    F: Future<Output = Result<Response<ResBody>, E>>,
{
    type Output = Result<Response<ResBody>, E>;

    fn poll(self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Self::Output> {
        let this = self.project();

        let mut response = ready!(this.inner.poll(cx))?;
        let mut cookie = Cookie::build(CSRF_COOKIE_NAME)
            .expires(Expiration::Session)
            .same_site(SameSite::Strict)
            .secure(true)
            .build();

        let guard = this.handle.inner.lock().unwrap();
        if let Some(ref set_data) = guard.set_data {
            let value = format!("{}.{}", set_data.hash, set_data.message);
            cookie.set_value(value);
        } else {
            cookie.make_removal();
        }

        let encoded_cookie = cookie.encoded().to_string();
        let header_value = HeaderValue::from_str(&encoded_cookie).unwrap();

        response
            .headers_mut()
            .append(header::SET_COOKIE, header_value);

        Poll::Ready(Ok(response))
    }
}
