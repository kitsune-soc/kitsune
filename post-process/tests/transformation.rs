use futures_util::future;
use post_process::{Element, Html, Transformer};
use pretty_assertions::assert_eq;
use std::borrow::Cow;

#[tokio::test]
async fn link_transformation() {
    let text = "@真島@goro.org how are you doing? :friday-night: #龍が如く0";
    let transformer = Transformer::new(|elem| async move {
        let transformed = match elem {
            Element::Emote(emote) => Element::Html(Html {
                tag: Cow::Borrowed("a"),
                attributes: vec![(
                    Cow::Borrowed("href"),
                    Cow::Owned(format!("https://example.com/emote/{}", emote.content)),
                )],
                content: Box::new(Element::Emote(emote)),
            }),
            Element::Hashtag(hashtag) => Element::Html(Html {
                tag: Cow::Borrowed("a"),
                attributes: vec![(
                    Cow::Borrowed("href"),
                    Cow::Owned(format!("https://example.com/hashtag/{}", hashtag.content)),
                )],
                content: Box::new(Element::Hashtag(hashtag)),
            }),
            Element::Mention(mention) => Element::Html(Html {
                tag: Cow::Borrowed("a"),
                attributes: vec![(
                    Cow::Borrowed("href"),
                    Cow::Owned(format!(
                        "https://example.com/mention/{}/{}",
                        mention.username,
                        mention.domain.as_deref().unwrap_or_default()
                    )),
                )],
                content: Box::new(Element::Mention(mention)),
            }),
            elem => elem,
        };

        Ok(transformed)
    });
    let transformed = transformer.transform(text).await.unwrap();

    assert_eq!(
        transformed,
        r#"<a href="https://example.com/mention/真島/goro.org">@真島@goro.org</a> how are you doing? <a href="https://example.com/emote/friday-night">:friday-night:</a> <a href="https://example.com/hashtag/龍が如く0">#龍が如く0</a>"#
    );
}

#[tokio::test]
async fn noop_transformation() {
    let text = "@真島@goro.org how are you doing? :friday-night: #龍が如く0";
    let transformer = Transformer::new(future::ok);
    let transformed = transformer.transform(text).await.unwrap();

    assert_eq!(text, transformed);
}
