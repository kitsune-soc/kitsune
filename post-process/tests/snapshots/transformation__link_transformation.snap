---
source: post-process/tests/transformation.rs
expression: transformer.transform(text).await.unwrap()
---
<a href="https://example.com/mention/真島/goro.org">@真島@goro.org</a> how are you doing? <a href="https://example.com/emote/friday-night">:friday-night:</a> <a href="https://example.com/hashtag/龍が如く0">#龍が如く0</a>
